### PR TITLE
Feat/add influxdb metrics provider

### PR DESCRIPTION
**Pull request by jaymebrd**
WIP

### CODE REVIEW

File: docs/analysis/influxdb.md [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+spec:
+  args:
+  - name: application-name
+  metrics:
+  - name: error-rate
+    # NOTE: To be consistent with the prometheus metrics provider InfluxDB query results are returned as an array.
+    # In the example we're looking at index 0 of the returned array to obtain the value we're using for the success condition
+    successCondition: result[0] <= 0.01
+    provider:
+      newRelic:
```
Comment (GeorgeMac) [inline docs/analysis/influxdb.md]: I may just be missing something, but should this be `influxdb`?

File: metricproviders/influxdb/influxdb.go [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+	"time"
+
+	"github.com/argoproj/argo-rollouts/pkg/apis/rollouts/v1alpha1"
+	metricutil "github.com/argoproj/argo-rollouts/utils/metric"
+)
+
+const (
+	//ProviderType indicates the provider is InfluxDB
+	ProviderType                    = "Influxdb"
+	DefaultInfluxdbTokensSecretName = "influxdb"
```
Comment (GeorgeMac) [inline metricproviders/influxdb/influxdb.go:24]: Maybe a comment here (I tend to comment on all exported types - bit of a hangover from over-zealous vetting tools).

Comment (GeorgeMac) [reply metricproviders/influxdb/influxdb.go:24]: Same goes for the one below 👇

File: metricproviders/influxdb/influxdb.go [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+}
+
+// Run queries influxdb for the metric
+func (p *Provider) Run(run *v1alpha1.AnalysisRun, metric v1alpha1.Metric) v1alpha1.Measurement {
+	startTime := metav1.Now()
+	newMeasurement := v1alpha1.Measurement{
+		StartedAt: &startTime,
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
```
Comment (GeorgeMac) [inline metricproviders/influxdb/influxdb.go]: Maybe pull this duration up into a constant:
```
const defaultQueryTimeout = 30 * time.Second
```

File: pkg/apis/rollouts/v1alpha1/generated.proto [hunk 4]
```diff
@@ -693,6 +693,15 @@ message GraphiteMetric {
   optional string query = 2;
 }
 
+// GraphiteMetric defines the Graphite query to perform canary analysis
```
Comment (GeorgeMac) [inline pkg/apis/rollouts/v1alpha1/generated.proto]:
```suggestion
// InfluxdbMetric defines the InfluxDB Flux query to perform canary analysis
```

File: utils/analysis/factory.go [hunk 5]
```diff
@@ -219,6 +219,12 @@ func ValidateMetric(metric v1alpha1.Metric) error {
 	if metric.Provider.CloudWatch != nil {
 		numProviders++
 	}
+	if metric.Provider.Graphite != nil {
+		numProviders++
+	}
```
Comment (GeorgeMac) [inline utils/analysis/factory.go:224]: Quick question, why does this introduce graphite?

**Review by GeorgeMac — CHANGES_REQUESTED** [review]
Couple discrepancies and minor style nits.
Otherwise, this is looking great.

File: docs/analysis/influxdb.md (hunk 1 above)
Comment (jaymebrd) [reply docs/analysis/influxdb.md]: Oh no, I just copied their docs example

File: utils/analysis/factory.go (hunk 5 above)
Comment (jaymebrd) [reply utils/analysis/factory.go:224]: Same as above - I fixed same graphite issues along the way.

File: pkg/apis/rollouts/v1alpha1/generated.proto (hunk 4 above)
Comment (jaymebrd) [reply pkg/apis/rollouts/v1alpha1/generated.proto]: This is actually expected, I fixed same graphite issues along the way.

Comment (jaymebrd) [reply pkg/apis/rollouts/v1alpha1/generated.proto]: Actually my mistake, the graphite stuff I added is elsewhere - nice catch!

File: utils/analysis/factory.go (hunk 5 above)
Comment (jaymebrd) [reply utils/analysis/factory.go:224]: Do you think I should drop the graphite fixup? I just noticed it was missing here.

Comment (GeorgeMac) [reply utils/analysis/factory.go:224]: Ahh good question. I'd say contribute it and let Argo team decided if they want to split it out.

**Review by GeorgeMac — APPROVED** [review]
Looks good and ready to PR on rollouts 👍

**Pull request closed without merging**